ETS backend

arkgil opened the pull request:
This PR introduces an implementation based on ETS. In fact, there are two implementations, the second one being a slight improvement over the first one.

### 1. Single ETS 

You can find this implementation in `Events.Impl.Ets` module.

Each handler is stored in a bag ETS table. The key is an event prefix the handler is attached to. All writes to a table go through a single Agent process to make sure that we don't get duplicate handler IDs.

Reads (`list_handlers_...`) on the ETS table are performed directly by the calling processes. When looking up handlers for event, first all the prefixes of event are built. Having those prefixes, a match spec is constructed to search for all handlers subscribed to these prefixes.
Surprisingly (or maybe not?) ETS is so fast, that the bottleneck of `list_handlers_for_event/1`
is the procedure of building all possible prefixes.

When it comes to concurrency guarantess, there are simply none. When one process performs
reads handlers for event, it might happen that another process deletes a handler from the table
(because `:ets.select/2` is used which is not atomic).

### 2. Single ETS + cache

You can find this implementation in `Events.Impl.EtsCached` module. It is based on two ETS tables: regular table for handlers as in `Ets` implementation, and a cache table.
                                                                                                  
If the cache is disabled (more on when it's enabled later) the reads works exactly like in the `Ets` implementation.
                                                                                                  
The cache is a set table with read concurrency. The key is an event name (not a prefix!) and
the value is a lists of handler records as stored in regular handler table. As you might suspect,
the list contains all handlers which should be invoked when the event stored as key is emitted.
                                                                                                  
When the cache is enabled, `list_handlers_for_event/1` first checks if the handlers which
should be invoked for this event are not already cached. If they are, then those handlers are
returned. If they are not, the handlers are looked up in regular table, exactly the same as
in `Ets` implementation, and then they are cached and returned from the function.
                                                                                                  
In this implementation Agent has been replaced with the GenServer. The same as in `Ets` backend,
all writes go though this process. However in this case, the GenServer is also responsible for
enabling and disabling the cache. The cache is enabled when there have been no writes for 500
milliseconds. When the cache is enabled and the write occurs, the whole cache is cleared and
the countdown starts again. The downside of this approach is that the cache will be cleared
when handlers are detached due to their failure.
The improvement here would be to delete only these cache entries which would be affected by the
write (fortunately this quite easy to achive and doesn't require computing the prefixes).
With such solution, the cache would be always enabled. It would probably be beneficial to
detach failing handlers in bathes to lower the number of writes to the table.
There is also another downside of this solution: all writes to cache table go through the GenServer.
It might happen that many processes emit the same event at once and all try to save handlers
in cache which might flood the process with messages. On the other hand, if the processes would
insert entries into cache by themselves, we could run into all sorts of concurrency issues.
                                                                                                  
One thing missing from this implementation is cache invalidation. I believe that the LFU (least
frequently used) scheme would be the most appropriate here, although some more elaborate techniques must be needed.

### Discussion

From the benchmarks I've run it turns out that ETS+cache is an order of magnitude faster (measured by iterations per second of `list_handlers_for_event/1` callback) than the bare ETS implementation. Also it looks like the Agent implementation is **faster** than the bare ETS. I've done some profiling and it happens most probably because the process of building all prefixes for the event is really expensive. Unfortunately, and quite expectedly, Agent implementation fails when there are more than 1000 processes running in the benchmark (with the default timeout of 5000ms).

Obviously the implementation with cache introduces all the cache-related problems, most notably how to evict records. I don't have more ideas on how to speed things up given the current API and semantics of attaching handlers to event prefixes. I know that @studzien has experimented with tree implementation, but as I understood it was still slower than the ETS+cache.

I can post the results of benchmarks if anyone's interested. You can also run them by yourselves using the `mix bench.load_handlers_for_event` task (`MIX_ENV=bench mix help bench.list_handlers_for_event` might help here 🙂 ).

/cc @josevalim @chrismccord @cloud8421 @studzien @mohamedalikhechine @rslota

josevalim commented:
Great job @arkgil!

What happens if you precompile the match spec in the first case? Imagine that you have a function called `compile_event_name!(...)` that emits the matchspec then all lookups will already have a precompiled matchspec. If this option is feasible, then we can figure out a mechanism for precompilation of event names (for example, we can do it at compile time).

josevalim commented on bench/list_handlers_for_event.ex:
```diff
@@ -0,0 +1,134 @@
+  @switches [
+    parallelism: :integer,
+    duration: :integer,
+    "handlers-count": :integer,
```
You can pass `:handlers_count` and the parser will figure it out.

arkgil replied:
Thanks! I didn't know about it, neat 😉

josevalim commented:
@arkgil it is also worth saying that because the id in the table is the `handler_id`, your selects end-up traversing the table. So the time to lookup handlers will be `NUMBER_OF_HANDLERS * NUMBER_OF_PREFIXES`. Plus, the number of prefixes is a  list with repeated prefixes. For example, if the event name is `[:foo, :bar, :baz]` we will compare against `[:foo]`, `[:foo, :bar]` and `[:foo, :bar, :baz]`.

Maybe a better design is to make the prefix be the key. This means the lookup will always be hash based. This means deletions are slower but that's a reasonable trade-off to make. It also means we need to change the table type to a `duplicate_bag`.

Also please make sure to measure both the cost of a hit and a miss. :)

arkgil commented:
@josevalim thanks for the response!

If you look at the definition of the handlers table:

```elixir
:ets.new(@table, [:bag, :protected, :named_table, keypos: 2, read_concurrency: true])
```

you'll notice that the prefix is already the key 🙂 unless I made some dumb mistake 😅 If I haven't then I think that the select is as fast as possible with this table structure, because each pattern in a match spec includes a key. It works with regular `bag` type because we guarantee each record being unique thanks to uniqueness of `handler_id`.

> Also please make sure to measure both the cost of a hit and a miss.

Yes, I need to prepare a benchmark for that too. I suspect the performance to be similar to that of regular ETS implementation, but it might be even worse because on each lookup we check if cache is enabled.

arkgil commented:
@josevalim precompiling the match spec is an awesome idea! I'll try it as soon as possible.

josevalim commented:
Oh, I definitely missed the key_pos, thanks! In this case, I am almost sure that ETS won't do a full table traversal. In this case, just for sanity check purposes, can you also benchmark us performing a `ets:lookup/3` per key instead of using a match spec?

arkgil commented:
@josevalim sure, good call :)

arkgil commented:
I've added the ETS backend which performs lookups instead of `:ets.select` created by @studzien .

Here are the results of 20 second benchmarks of `list_handlers_for_event/1` with 100 handlers in table and 8 concurrent processes (as many as schedulers). Each call in benchmark is a cache hit in case of cache backend. Performance of `lookup` surprisingly looks much better than performance of `select`:

```
Name                      ips        average  deviation         median         99th %
ETS with cache        22.36 K       44.72 μs   ±209.94%          32 μs         132 μs
ETS with lookup        4.40 K      227.46 μs    ±68.83%         216 μs         455 μs
Agent                  1.93 K      518.47 μs    ±46.48%         479 μs        1323 μs
ETS                    0.61 K     1648.88 μs    ±50.75%        1504 μs        4925 μs

Comparison:
ETS with cache        22.36 K
ETS with lookup        4.40 K - 5.09x slower
Agent                  1.93 K - 11.59x slower
ETS                    0.61 K - 36.87x slower
```

I'd even say that this performance is acceptable given that this benchmark is biased towards cache backend (because every call is a cache hit).

(But please note that implementation with cache still uses `:ets.select`).

I know that @studzien also experimented with generating prefixes at compile time which gave ~20% improvement to ips of ETS with lookup. The downside of this approach is that we need to know literal event name at compile time. Maybe the solution could be to generate prefixes at compile time when possible and at runtime when not?

In any case, for now I'd leave the backend with lookups as **the** ETS implementation. We can try to improve it further one we have more realistic benchmarking scenarios.

cloud8421 commented:
@arkgil said:

> I know that @studzien also experimented with generating prefixes at compile time which gave ~20% improvement to ips of ETS with lookup. The downside of this approach is that we need to know literal event name at compile time. Maybe the solution could be to generate prefixes at compile time when possible and at runtime when not?

I would agree to optimize at compile time when possible provided that it doesn't become a complication which can be avoided for a first release (being aware of the performance impact seems enough for now).

josevalim commented:
@arkgil can you please change the table to `duplicate_bag`? It should not affect benchmarks but it should improve insertion times.

studzien commented:
I think it would be good to first try to use it in some real project (like Phoenix) and this way gather data for a more realistic benchmark, rather than optimize prematurely :)

josevalim commented:
To be honest, I am not glad with those numbers. It means that 5 measurements will add a cost of 1ms. Although the number of 100 handlers is probably unlikely. What happens with 10 handlers?

josevalim commented:
@studzien oh yeah, that looks much better, thanks! :heart: Can we limit the size of the event to 5 and attach 100 handlers? I think a depth of 5 is much more realistic, although we will probably stay at 2 or 3.

And sorry for asking instead of measuring it myself but I can't jump into the repo right now.

studzien commented:
I added an option to control number of matching handlers in the benchmark (`-m`), so you can play with this.

For future reference (`-h` is the number of handlers attached, `-m` is the number of returned handlers):

```
$ mix bench.list_handlers_for_event -h 100 -m 100 -d 20
Name                      ips        average  deviation         median         99th %
ETS with cache        22.32 K       44.81 μs   ±220.22%          32 μs         141 μs
ETS with lookup        4.54 K      220.38 μs    ±93.82%         209 μs         427 μs
Agent                  3.02 K      331.51 μs    ±34.04%         319 μs         600 μs
ETS                    0.59 K     1691.35 μs    ±33.62%        1646 μs     3960.10 μs

Comparison:
ETS with cache        22.32 K
ETS with lookup        4.54 K - 4.92x slower
Agent                  3.02 K - 7.40x slower
ETS                    0.59 K - 37.75x slower
```

```
$ mix bench.list_handlers_for_event -h 10 -m 10 -d 20
Name                      ips        average  deviation         median         99th %
ETS with cache       498.88 K        2.01 μs  ±3593.34%           2 μs           2 μs
ETS with lookup      142.40 K        7.02 μs   ±755.82%           6 μs          13 μs
ETS                   66.72 K       14.99 μs   ±344.83%          14 μs          21 μs
Agent                 53.19 K       18.80 μs   ±302.69%          15 μs          67 μs

Comparison:
ETS with cache       498.88 K
ETS with lookup      142.40 K - 3.50x slower
ETS                   66.72 K - 7.48x slower
Agent                 53.19 K - 9.38x slower
```

```
$ mix bench.list_handlers_for_event -h 100 -m 5 -d 20
Name                      ips        average  deviation         median         99th %
ETS with cache       702.14 K        1.42 μs  ±4364.42%           1 μs           4 μs
ETS with lookup      290.23 K        3.45 μs  ±1555.15%           3 μs           5 μs
ETS                  139.29 K        7.18 μs   ±649.69%           7 μs           9 μs
Agent                  5.25 K      190.56 μs    ±61.71%         173 μs         422 μs

Comparison:
ETS with cache       702.14 K
ETS with lookup      290.23 K - 2.42x slower
ETS                  139.29 K - 5.04x slower
Agent                  5.25 K - 133.80x slower
```

```
$ mix bench.list_handlers_for_event -h 1000 -m 5 -d 20
Name                      ips        average  deviation         median         99th %
ETS with cache       634.58 K        1.58 μs  ±4822.62%           1 μs        2.20 μs
ETS with lookup      285.23 K        3.51 μs  ±1566.45%           3 μs           5 μs
ETS                   98.24 K       10.18 μs  ±1216.56%           8 μs          17 μs
Agent               0.00708 K   141293.39 μs    ±34.05%   132016.50 μs   320592.43 μs

Comparison:
ETS with cache       634.58 K
ETS with lookup      285.23 K - 2.22x slower
ETS                   98.24 K - 6.46x slower
Agent               0.00708 K - 89661.22x slower
```

I'm going to try some bigger handler numbers as well, but it seems that the attachment phase needs some speedup first, so that will also check difference between `bag` and `duplicate_bag`.

josevalim commented:
The bag vs duplicated_bag is not supposed to impact performance. Basically bag checks for any matching object under the same key, which is not necessary since we already do a check for the handler_id. Using bag instead duplicate_bag was what cause one of the slow downs in the first version of Phoenix PubSub.

Also, how would someone subscribe to **all** events?

josevalim commented:
@studzien and thanks for the `-m` option, those numbers look much more acceptable now. I think we can ship it, once we figure out the ets table type and how to subscribe to all events. Unless we don't want to allow subscribing to all.

arkgil commented:
Thanks everyone for the input! ❤️ These numbers look great now! 😄 

@josevalim 
> Also, how would someone subscribe to all events?

You need to attach handler to empty list since it's a prefix to all event names:

```elixir
Events.attach("id", [], Handler, :handle)
```

arkgil commented:
Oh, and just to make sure we're all on the same page - in benchmarks we measure the time of performing lookup. There will be additional cost of going through all the handlers and invoking them.

arkgil commented:
@josevalim I'll add a benchmark suite for that.

arkgil commented:
It looks like invoking handlers doesn't add much overhead. 

It's also worth noting that the `-m` option determines the length of the event we're invoking. So it might happen that the speedup we see when we lower this number comes from the fact that there are fewer prefixes to build. Let's keep in the backs of our heads that we can still play with prefix precompilation if things will become too slow.

josevalim commented:
Yup, this looks good to go to me then! We can ship the ETS lookup and add a cache later if we want to.

arkgil commented:
If no one objects until tomorrow I'll clean this up and we'll merge! :)

arkgil commented:
I've renamed the `EtsLookup` module to `Ets` and removed other ETS implementations. We can always recover them from the git history if necessary. I think it's ready for merge.

josevalim commented:
@arkgil should we get rid of the "Impl" indirection as well and just inline the code into the API?

arkgil commented:
@josevalim I'm afraid that we might forget about the module boundaries. I'd leave the behaviour module for now, just in case we change the implementation before first release.

josevalim commented:
@arkgil sounds good to me, your call :)

josevalim commented:
@arkgil note we are having a discussion on the ecto mailing list and I mentioned this library: https://groups.google.com/forum/#!topic/elixir-ecto/F0SJKj2Mfak

And @GregMefford suggested something that I agree with: what if we start with just a single ets lookup for now without the prefixes? Then we keep this as simple and as fast as it can be and then, based on the use cases, we can find out how we want to move to other approaches.

We already have the benchmarks, we have validated them and we know it is doable but we don't have the use cases. So we can wait for the use cases and add it back only if necessary. Thoughts?

studzien commented:
@josevalim Would you mind moving this discussion to a separate issue? Perhaps in the new repo already :)

arkgil commented:
@josevalim I'll create an issue on Telemetry repository for further discussion.
